src/protocol/sort/multi_bit_permutation.rs

```diff
@@ -61,18 +61,22 @@ pub async fn multi_bit_permutation<'a, F: Field, S: SecretSharing<F>, C: Context
     }
 
     // Take sum of products of output of equality check and accumulated sum
-    let mut one_off_permutation =
-        try_join_all((0..num_records).zip(repeat(ctx)).map(|(rec, ctx)| {
-            let mut sop_inputs = Vec::with_capacity(num_possible_bit_values);
-            for idx in 0..num_possible_bit_values {
-                sop_inputs.push((&equality_checks[rec][idx], &prefix_sum[rec][idx]));
-            }
-            async move {
-                ctx.sum_of_products(RecordId::from(rec), sop_inputs.as_slice())
-                    .await
-            }
-        }))
-        .await?;
+    let mut one_off_permutation = try_join_all(
+        equality_checks
+            .into_iter()
+            .zip(prefix_sum.into_iter())
+            .zip(repeat(ctx))
+            .enumerate()
+            .map(|(i, ((eq_checks, prefix_sums), ctx))| async move {
+                ctx.sum_of_products(
+                    RecordId::from(i),
+                    eq_checks.iter().collect::<Vec<_>>().as_slice(),
+                    prefix_sums.iter().collect::<Vec<_>>().as_slice(),
```

Comment:
This is stupid.

This is here because `sum_of_products` really wants to accept a slice of *references* to shares... what we have is vectors of *shares*. So this converts them into vectors of *references*...

I attempted to eliminate this, and make `sum_of_products` just accept a slice of shares... but unfortunately, while this is easy for the semi-honest one, this isn't easy for the malicious version. That's because it attempts to call the semi-honest version, but it doesn't *have* vectors of replicated shares - it has vectors of *malicious replicated*...

I couldn't fix this without copy-pasting a lot of code. We can return to this later. There is another optimisation waiting here.

Reply:
Lets open an issue for this so we address it soon

Reply:
#403